### PR TITLE
feat: challenge helpers for non-step-based challenges

### PR DESCRIPTION
**Pull request by naomi-lgbt**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

With work being done on things like the Odin Project, not all of our upcoming lessons will be step-based. This adds tools similar to the step-based tools for adding and manipulating the challenges and their order.

### CODE REVIEW

File: tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { prompt } from 'inquirer';
+import { challengeTypes } from '../../../client/utils/challenge-types';
+
+export const newChallengePrompts = async (): Promise<{
+  name: string;
+  title: string;
+  challengeType: string;
+}> => {
+  const title = await prompt({
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts:9]:
```suggestion
  const dashedName = await prompt({
```
'name' and 'title' are back to front, also name is a bit too generic imo

File: tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      if (/[^a-z0-9-]/.test(block)) {
+        return 'please use alphanumerical characters and kebab case';
+      }
+      return true;
+    },
+    filter: (block: string) => {
+      return block.toLowerCase();
+    }
+  });
+  const name = await prompt({
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts:25]:
```suggestion
  const title = await prompt({
```

File: tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    type: 'list',
+    choices: Object.entries(challengeTypes).map(([key, value]) => ({
+      name: key,
+      value
+    }))
+  })) as { value: string };
+
+  return {
+    title: String(title.value),
+    name: String(name.value),
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts:42]:
```suggestion
    dashedName: String(dashedName.value),
```

File: tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts (hunk 1 above)
Comment (ojeytonwilliams) [reply tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts:9]: Also, `prompt` is generic, so you can do something like 

```ts
  const { dashedName } = await prompt<{ dashedName: string }>({
    name: 'dashedName',
```
which is a little less awkward than either using `String` or `as { value: string }` to mollify TS.

File: tools/challenge-helper-scripts/insert-challenge.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  const challengeAfter = (await prompt({
+    name: 'id',
+    message: 'Which challenge should come AFTER this new one?',
+    type: 'list',
+    choices: challenges.map(([id, title]) => ({
+      name: title,
+      value: id
+    }))
+  })) as { id: string };
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/insert-challenge.ts:25]:
```suggestion
  const challengeAfter = await prompt<{id: string}>({
    name: 'id',
    message: 'Which challenge should come AFTER this new one?',
    type: 'list',
    choices: challenges.map(([id, title]) => ({
      name: title,
      value: id
    }))
  })
```
There's not much in it, but this seems a little cleaner.

File: tools/challenge-helper-scripts/insert-challenge.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  const challengeId = new ObjectID();
+  const template = templateGenerator({ ...options, challengeId });
+  createChallengeFile(options.title, template, path);
+
+  const meta = getMetaData();
+  meta.challengeOrder.splice(indexToInsert, 0, [
+    challengeId.toString(),
+    options.title
+  ]);
+  meta.challengeOrder.push([challengeId.toString(), options.title]);
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/insert-challenge.ts:43]:
```suggestion
```
Only need the one.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: tools/challenge-helper-scripts/insert-challenge.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    ([id]) => id === challengeAfter.id
+  );
+
+  const templateGenerator = challengeTypeToTemplate[options.challengeType];
+  if (!templateGenerator) {
+    return;
+  }
+  const challengeId = new ObjectID();
+  const template = templateGenerator({ ...options, challengeId });
+  createChallengeFile(options.title, template, path);
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/insert-challenge.ts:36]:
```suggestion
  createChallengeFile(options.dashedName, template, path);
```

File: tools/challenge-helper-scripts/create-next-challenge.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  const path = getProjectPath();
+
+  const options = await newChallengePrompts();
+  const templateGenerator = challengeTypeToTemplate[options.challengeType];
+  if (!templateGenerator) {
+    return;
+  }
+  const challengeId = new ObjectID();
+  const template = templateGenerator({ ...options, challengeId });
+  createChallengeFile(options.title, template, path);
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/create-next-challenge.ts:18]:
```suggestion
  createChallengeFile(options.dashedName, template, path);
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Ooops, missed these.

File: tools/challenge-helper-scripts/insert-challenge.ts (hunk 5 above)
Comment (naomi-lgbt) [reply tools/challenge-helper-scripts/insert-challenge.ts:43]: Whoops

**Comment by ojeytonwilliams** [comment]
Sorry, Naomi, I didn't do a good job with those suggestions. I'll push a cleanup commit in the morning.

File: tools/challenge-helper-scripts/delete-challenge.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  await unlink(`${path}${fileToDelete}`);
+
+  const meta = getMetaData();
+  meta.challengeOrder.splice(indexToDelete, 1);
+  updateMetaData(meta);
+};
+
+void (async () => await deleteChallenge())();
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/delete-challenge.ts:38]:
```suggestion
void createNextChallenge();
```

File: tools/challenge-helper-scripts/insert-challenge.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  const meta = getMetaData();
+  meta.challengeOrder.splice(indexToInsert, 0, [
+    challengeId.toString(),
+    options.title
+  ]);
+  updateMetaData(meta);
+};
+
+void (async () => await insertChallenge())();
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/insert-challenge.ts:46]:
```suggestion
void insertChallenge();
```

File: tools/challenge-helper-scripts/utils.ts [hunk 10]
```diff
@@ -31,6 +31,14 @@ const createStepFile = ({
   return challengeId;
 };
 
+const createChallengeFile = (
+  title: string,
+  template: string,
+  path = getProjectPath()
+): void => {
+  fs.writeFileSync(`${path}${title}.md`, template);
```
Comment (ojeytonwilliams) [inline tools/challenge-helper-scripts/utils.ts:39]: Not a big deal, but there's nothing that needs to wait for this, so it could be async.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
It's not really relevant to this PR, but I wonder if we should switch the meta.json format from

```
  "challengeOrder": [
    [
      "id": "587d7dbc367417b2b2512bb1",
      "title": "Create a Simple JSX Element"
    [,
```
to
```
  "challengeOrder": [
    {
      "id": "587d7dbc367417b2b2512bb1",
      "title": "Create a Simple JSX Element"
    },
```

Working with tuples is nasty and, weirdly, I don't think much code needs to change to accommodate this. What do you reckon?

Anyways, back on topic, some nitpicks:

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

<sub>IIFE are fine - I don't like _unnecessary_ ones!</sub>

File: docs/how-to-work-on-coding-challenges.md [hunk 11]
```diff
@@ -500,3 +500,66 @@ Creating and Editing Challenges:
 1. [Challenge types](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/utils/challenge-types.js#L1-L13) - what the numeric challenge type values mean (enum).
 
 2. [Contributing to FreeCodeCamp - Writing ES6 Challenge Tests](https://www.youtube.com/watch?v=iOdD84OSfAE#t=2h49m55s) - a video following [Ethan Arrowood](https://twitter.com/ArrowoodTech) as he contributes to the old version of the curriculum.
+
+## Helper Scripts
+
+There are a few helper scripts that can be used to manage the challenges in a block. Note that these commands should all be run in the block directory. For example:
```
Comment (moT01) [inline docs/how-to-work-on-coding-challenges.md:506]:
```suggestion
> [!NOTE]
> If you want to work with the step-based challenges, refer to the [Work on Practice Projects](how-to-work-on-practice-projects.md) section.

There are a few helper scripts that can be used to manage the challenges in a block. Note that these commands should all be run in the block directory. For example:
```

**Review by moT01 — COMMENTED** [review]
These tools are fantastic @naomi-lgbt 🎉 I left one suggestion to try and minimize confusion. Not sure if I worded it the best - but if someone sees this, they may not know about the other tools without a mention of them.

I would be fine adding this without the mention though, so let us know if you are ready to get it in.

**Review by moT01 — APPROVED** [review]

**Comment by lloydchang** [comment]
> It's not really relevant to this PR, but I wonder if we should switch the meta.json format from
> 
> ```
>   "challengeOrder": [
>     [
>       "id": "587d7dbc367417b2b2512bb1",
>       "title": "Create a Simple JSX Element"
>     [,
> ```
> 
> to
> 
> ```
>   "challengeOrder": [
>     {
>       "id": "587d7dbc367417b2b2512bb1",
>       "title": "Create a Simple JSX Element"
>     },
> ```
> 
> Working with tuples is nasty and, weirdly, I don't think much code needs to change to accommodate this. What do you reckon?
> 
> Anyways, back on topic, some nitpicks:

I believe this relates to https://github.com/freeCodeCamp/classroom/issues/410#issuecomment-1657303141

Context: [feat: Detect changes coming from FCC API responses inside of the classroom mode project #410](https://github.com/freeCodeCamp/classroom/issues/410)

**Pull request merged**